e2e/rbd.go

```diff
@@ -349,6 +350,228 @@ var _ = Describe("RBD", func() {
+				}
+				err = createRBDStorageClass(f.ClientSet, f, nil, nil, deletePolicy)
+				if err != nil {
+					e2elog.Failf("failed to create storageclass with error %v", err)
+				}
+			})
+
+			By("perform IO on rbd-nbd volume after nodeplugin restart and expect a failure", func() {
+				err := deleteResource(rbdExamplePath + "storageclass.yaml")
+				if err != nil {
+					e2elog.Failf("failed to delete storageclass with error %v", err)
+				}
+				// Storage class with rbd-nbd mounter
+				err = createRBDStorageClass(f.ClientSet, f, nil, map[string]string{"mounter": "rbd-nbd"}, deletePolicy)
+				if err != nil {
+					e2elog.Failf("failed to create storageclass with error %v", err)
+				}
+				pvc, err := loadPVC(pvcPath)
+				if err != nil {
+					e2elog.Failf("failed to load PVC with error %v", err)
+				}
+				pvc.Namespace = f.UniqueName
+
+				app, err := loadApp(appPath)
+				if err != nil {
+					e2elog.Failf("failed to  load application with error %v", err)
+				}
+
+				app.Namespace = f.UniqueName
+				label := map[string]string{
+					"app": app.Name,
+				}
+				app.Labels = label
+				app.Spec.Volumes[0].PersistentVolumeClaim.ClaimName = pvc.Name
+				err = createPVCAndApp("", f, pvc, app, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to create PVC and application with error %v", err)
+				}
+
+				// validate created backend rbd images
+				validateRBDImageCount(f, 1)
+
+				selector, err := getDaemonSetLabelSelector(f, cephCSINamespace, rbdDaemonsetName)
+				if err != nil {
+					e2elog.Failf("failed to get the labels with error %v", err)
+				}
+				// delete rbd nodeplugin pods
+				err = deletePodWithLabel(selector, cephCSINamespace, false)
+				if err != nil {
+					e2elog.Failf("fail to delete pod with error %v", err)
+				}
+
+				// wait for nodeplugin pods to come up
+				err = waitForDaemonSets(rbdDaemonsetName, cephCSINamespace, f.ClientSet, deployTimeout)
+				if err != nil {
+					e2elog.Failf("timeout waiting for daemonset pods with error %v", err)
+				}
+
+				opt := metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("app=%s", app.Name),
+				}
+				// Write something to mountpoint and expect a failure
+				filePath := app.Spec.Containers[0].VolumeMounts[0].MountPath + "/test"
+				_, stdErr := execCommandInPodAndAllowFail(f, fmt.Sprintf("echo 'Hello World' > %s", filePath), app.Namespace, &opt)
+				IOErr := fmt.Sprintf("cannot create %s: Input/output error", filePath)
+				if !strings.Contains(stdErr, IOErr) {
+					e2elog.Failf(stdErr)
+				} else {
+					e2elog.Logf("failed IO as expected: %v", stdErr)
+				}
+
+				err = deletePVCAndApp("", f, pvc, app)
+				if err != nil {
+					e2elog.Failf("failed to delete PVC and application with error %v", err)
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 0)
+			})
+
+			By("restart rbd-nbd process on nodeplugin and continue IO after nodeplugin restart", func() {
```

Comment:
Just a question Is this whole test case required? as we are doing some hacks to remount the rbd image which is not the case in the real world and this is expected to handled automatically by cephcsi or other sidecars in daemonset pod?

Reply:
currently this PR adds two tests  `perform IO on rbd-nbd volume after nodeplugin restart and expect a failure` and `restart rbd-nbd process on nodeplugin and continue IO after nodeplugin restart`  but when we bring the automatic handling of the restart, one test will be removed and other will be modified to check we are able to read/write the data? IMO having single test seems fine for me. Lets hear from others

Reply:
These are the POC level tests, for someone who is new to rbd-nbd mounter can understand the problem we are dealing here very clearly with the help of these test cases.

In the future when the real story is working we can clear these bits and makeup more useful test cases.

Till then the intention of these tests is to catch some regressions or failures at rbd-nbd only.